packages/zoe/src/zoeService/installationStorage.js

```diff
@@ -3,49 +3,82 @@
 import { assert, details as X } from '@agoric/assert';
 import { Far } from '@endo/marshal';
 import { E } from '@agoric/eventual-send';
+import { makeWeakStore } from '@agoric/store';
+
+/** @typedef { import('@agoric/swingset-vat').BundleID} BundleID */
 
 /**
- *
+ * @param {GetBundlecapForID} getBundlecapForID
  */
-export const makeInstallationStorage = () => {
-  /** @type {WeakSet<Installation>} */
-  const installations = new WeakSet();
+export const makeInstallationStorage = getBundlecapForID => {
+  /** @type {WeakStore<Installation, { bundlecap: Bundlecap, bundleID: BundleID }>} */
+  const installationsBundlecap = makeWeakStore('installationsBundlecap');
+  /** @type {WeakStore<Installation, SourceBundle>} */
+  const installationsBundle = makeWeakStore('installationsBundle');
 
   /**
-   * Create an installation by permanently storing the bundle. The code is
-   * currently evaluated each time it is used to make a new instance of a
-   * contract. When SwingSet supports zygotes, the code will be evaluated once
-   * when creating a zcfZygote, then the start() function will be called each
-   * time an instance is started.
+   * Create an installation from a bundle ID or a full bundle. If we are
+   * given a bundle ID, wait for the corresponding code bundle to be received
+   * by the swingset kernel, then store its bundlecap. The code is currently
+   * evaluated each time it is used to make a new instance of a contract.
+   * When SwingSet supports zygotes, the code will be evaluated once when
+   * creating a zcfZygote, then the start() function will be called each time
+   * an instance is started.
    */
-  /** @type {Install} */
-  const install = async bundle => {
-    assert.typeof(bundle, 'object', X`a bundle must be provided`);
+
+  /** @type {InstallBundle} */
+  const installBundle = async bundle => {
+    assert.typeof(bundle, 'object', 'a bundle must be provided');
     /** @type {Installation} */
     const installation = Far('Installation', {
       getBundle: () => bundle,
     });
-    installations.add(installation);
+    installationsBundle.init(installation, bundle);
     return installation;
   };
 
-  const assertInstallation = installation =>
-    assert(
-      installations.has(installation),
-      X`${installation} was not a valid installation`,
-    );
+  /** @type {InstallBundleID} */
+  const installBundleID = async bundleID => {
+    assert.typeof(bundleID, 'string', `a bundle ID must be provided`);
+    // this waits until someone tells the host application to store the
+    // bundle into the kernel, with controller.validateAndInstallBundle()
+    const bundlecap = await getBundlecapForID(bundleID);
```

Comment:
I'm happy to use `.then` here instead. Does your bias sort of feel that `await` is for things that will happen soon, whereas `.then` is for things that happen much later?

I don't think either form is better than the other w.r.t. upgrade or snapshots. Having a name for the second half helps to clarify its purpose, sometimes, but mostly when there might be another code path that leads into it (as in your comments in #4545 and the three different ways to obtain a promise for the ZCF bundle).

One style question: making both `installBundle` and `installBundleID` be `async` seems like a good idea to emphasize their parallelism (also they're both mostly invoked from remote vats, `E(zoe).install*`, so it's not like the caller gets a choice). Once `installBundleID` is `async`, then it seemed awkward to build up the `.then` and return a Promise instead of using `await` inline. If I change the `await` to `.then`, should I also change `installBundleID` to not be `async`? And then should I change `installBundle` to also not be async, and be ok with `installBundle` being actually-sync (and not returning a Promise) but `installBundleID` always returns a Promise? And, now they have different error behaviors if you e.g. pass in a bad argument (synchronous throw vs promise rejection).

Reply:
Please do *not* use `.then`. It is not safe against reentrancy. And least `await` is safe, and so switching from `await` to `.then` is a decrease in robustness. When you would use `.then`, use `E.when` instead.

If `await` and `E.when` are otherwise tied, a tie breaker is that `E.when` gives us deep stacks on errors. We cannot reasonably make that happen for `.then` or `await` without changes to the underlying engine.

Reply:
I do not consider it awkward for an `async` function to use explicit promise logic rather than `await`. In fact, I think we should do exactly this more often. In the absence of `await`, as you note, the meaning of `async` is to promise-ify the outcome, turning both success and errors into promises.

I will admit that this particular combination is one that other developers may find surprising when they first see it. In this case, I don't think that's a big hurdle.

Reply:
Note to folks following along at home (because I had to ask too): the concern with `.then` is a hostile provider of the object that you assume behaves like a Promise. It might be a non-Promise with a `.then` method that does something synchronously. It might be a real Promise but with an overriden `.then` method. Using `await` and `E.when` protects against this malice.

In this case, `getBundleCapForID` is coming from the rest of Zoe, but I'm happy to stick with something that maintains caution in any situation.

I'll stick with `await`, add an `// AWAIT` comment, and leave both functions as `async`.